### ISSUE TITLE
LSP doesn't handle server crashes well.

### ISSUE THREAD

**Issue by deathaxe** [tomv564/LSP#236]
At some point something got the `pyls` to crash and exit. LSP should handle the `BrokenPipeError` to either disable itself or try to restart the server as this is a kind of exception which you can't recover from in any other way.

At the moment the console is flooded with tracebacks while continuing working.

```
Failure writing payload
Traceback (most recent call last):
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 71, in send_payload
    self.process.stdin.flush()
BrokenPipeError: [Errno 32] Broken pipe

Exception in thread Thread-7:
Traceback (most recent call last):
  File "./python3.3/threading.py", line 901, in _bootstrap_inner
  File "./python3.3/threading.py", line 858, in run
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 91, in read_stdout
    header = self.process.stdout.readline()
AttributeError: 'NoneType' object has no attribute 'stdout'

Traceback (most recent call last):
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 110, in <lambda>
    lambda: purge_did_change(buffer_id, buffer_version), 500)
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 121, in purge_did_change
    notify_did_change(pending_buffer["view"])
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 195, in notify_did_change
    client.send_notification(Notification.didChange(params))
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 62, in send_notification
    self.send_payload(notification.to_payload())
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 70, in send_payload
    self.process.stdin.write(bytes(message, 'UTF-8'))
AttributeError: 'NoneType' object has no attribute 'stdin'
Traceback (most recent call last):
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 110, in <lambda>
    lambda: purge_did_change(buffer_id, buffer_version), 500)
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 121, in purge_did_change
    notify_did_change(pending_buffer["view"])
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 195, in notify_did_change
    client.send_notification(Notification.didChange(params))
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 62, in send_notification
    self.send_payload(notification.to_payload())
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 70, in send_payload
    self.process.stdin.write(bytes(message, 'UTF-8'))
AttributeError: 'NoneType' object has no attribute 'stdin'
Traceback (most recent call last):
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 110, in <lambda>
    lambda: purge_did_change(buffer_id, buffer_version), 500)
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 121, in purge_did_change
    notify_did_change(pending_buffer["view"])
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 195, in notify_did_change
    client.send_notification(Notification.didChange(params))
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 62, in send_notification
    self.send_payload(notification.to_payload())
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 70, in send_payload
    self.process.stdin.write(bytes(message, 'UTF-8'))
AttributeError: 'NoneType' object has no attribute 'stdin'
Traceback (most recent call last):
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 110, in <lambda>
    lambda: purge_did_change(buffer_id, buffer_version), 500)
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 121, in purge_did_change
    notify_did_change(pending_buffer["view"])
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 195, in notify_did_change
    client.send_notification(Notification.didChange(params))
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 62, in send_notification
    self.send_payload(notification.to_payload())
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 70, in send_payload
    self.process.stdin.write(bytes(message, 'UTF-8'))
AttributeError: 'NoneType' object has no attribute 'stdin'
Traceback (most recent call last):
  File "C:\Apps\Sublime\sublime_plugin.py", line 417, in run_async_view_listener_callback
    vel.__class__.__dict__[name](vel)
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\signature_help.py", line 66, in on_modified_async
    purge_did_change(self.view.buffer_id())
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 121, in purge_did_change
    notify_did_change(pending_buffer["view"])
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 195, in notify_did_change
    client.send_notification(Notification.didChange(params))
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 62, in send_notification
    self.send_payload(notification.to_payload())
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 70, in send_payload
    self.process.stdin.write(bytes(message, 'UTF-8'))
AttributeError: 'NoneType' object has no attribute 'stdin'
Traceback (most recent call last):
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 110, in <lambda>
    lambda: purge_did_change(buffer_id, buffer_version), 500)
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 121, in purge_did_change
    notify_did_change(pending_buffer["view"])
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\documents.py", line 195, in notify_did_change
    client.send_notification(Notification.didChange(params))
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 62, in send_notification
    self.send_payload(notification.to_payload())
  File "C:\Apps\Sublime\Data\Packages\LSP\plugin\core\rpc.py", line 70, in send_payload
    self.process.stdin.write(bytes(message, 'UTF-8'))
AttributeError: 'NoneType' object has no attribute 'stdin'
```

**Comment by tomv564** [comment]
I merged #238 but didn't take an extraordinary effort to test the behaviour. Please provide feedback in this issue if you're running from master with a crashy language server.

**Comment by tomv564** [comment]
Nice, thanks for the extensive testing! Good to know about OSError should it turn up in a user's log in the future.
